### PR TITLE
Added support for mobile sms

### PR DESCRIPTION
**Pull request by nishikeshKardak**
Using way2sms API we can send mobile sms. We need way2sms acoount.

### CODE REVIEW

**Comment by jarifibrahim** [comment]
Hey @nishikeshKardak :wave: 
Thank you so much for sending the PR. It looks good. :smiley: 

Can you please add some information to the PR description (and readme.md file) about the way2sms API and how I can test your PR?

For instance, I would like to know
1. What changed I need to make to test the PR
2. How to get the way2sms API key (if there is one)
3. What configuration parameters I need to pass so that I get the sms feature works.

**Comment by nishikeshKardak** [comment]
WAY2SMS provides good API for sending sms.
Both the apiKey and secretKey are available at way2sms user account.

capture4654

**Comment by nishikeshKardak** [comment]
I made prepare_sms() method in general.py file, it prepares message to send. In prepare_sms() method send_sms() gets called with message as a parameter. send_sms() takes message as argument and called sendPostRequest() method with apiKey, secretKey, phoneNo and message.

File: send_sms.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import requests
+import json
+import general
+
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
```
Comment (jarifibrahim) [inline send_sms.py]: It would make sense to move this (and other configs like apikey, secret key) to the config file https://github.com/jarifibrahim/resultBot/blob/master/config.py

File: main.py [hunk 2]
```diff
@@ -47,22 +47,21 @@ def get_files(files, dl_file_name):
...
             continue
         new_files.append(download_file(file, dl_file_name))
     return new_files
 
 
 def setup():
-    print '\n' + str(datetime.now().strftime('%Y-%m-%d %H:%M:%S'))
-    print '=' * 40
-    print "Fetching results related to " + FACULTY + " faculty"
+    print ('\n' + str(datetime.now().strftime('%d-%b-%Y %H:%M:%S')))
```
Comment (jarifibrahim) [inline main.py:61]: Thanks for fixing this.
`30-Jan-2019 12:17:02'` looks much better than `2019-01-30 12:19:30` :)

File: send_sms.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+
+# get request
+def sendPostRequest(reqUrl, apiKey, secretKey, useType, phoneNo, senderId, textMessage):
```
Comment (jarifibrahim) [inline send_sms.py]: Most of the params for this method should be pulled in from the config file. 
Can we do something similar to https://github.com/jarifibrahim/resultBot/blob/master/send_email.py#L7 ?

Comment (jarifibrahim) [inline send_sms.py]: Can you please change this to `send_post_request`? 
I've tried to follow the PEP8 naming conventions. See https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

File: send_sms.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import requests
+import json
+import general
+
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
+
+
+# get request
```
Comment (jarifibrahim) [inline send_sms.py]: The comment seems to be off :)

**Review by jarifibrahim — CHANGES_REQUESTED** [review]
Hey @nishikeshKardak most of the stuff looks good. I've added a few comments. Feel free to ask for help :)

File: general.py [hunk 5]
```diff
@@ -1,22 +1,23 @@
 ''' Some utility functions '''
 
 from bs4 import BeautifulSoup
-from urlparse import urljoin
+from urllib.parse import urlparse
```
Comment (jarifibrahim) [inline general.py]: Can you undo this change, please?
The project is intended to be used with python 2.7 and python 2.7 doesn't have a `urllib.parse` package.

File: main.py (hunk 2 above)
Comment (nishikeshKardak) [reply main.py:61]: Thanks for appreciating 😀

File: send_sms.py (hunk 1 above)
Comment (nishikeshKardak) [reply send_sms.py]: Hey @jarifibrahim 👋
Thanks for reveiwing PL
As per your requested changes, I did all required changes.

**Comment by nishikeshKardak** [comment]
Hey @jarifibrahim I resolve all the problems. Can you please review PL.

File: send_sms.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+def send_sms(message):
+    try:
+        # getting response and save as response object
+        response = send_post_request(URL, APIKEY, SECRETKEY, 'stage', PHONENO, 'WAYSMS', message )
+        # print response if you want
+        print (response.text)
+    except Exception as e:
+        print(e)
+    finally :
+        print()
```
Comment (jarifibrahim) [inline send_sms.py]: Can we remove this, please?

File: send_sms.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        exit()
```
Comment (jarifibrahim) [inline send_sms.py]: We might not want to exit if we couldn't send the SMS. For instance, I was running this script on an AWS instance 24x7. If the SMS failed, it would've stopped the program. This is what we would want.

File: send_sms.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+
+def send_sms(message):
+    try:
+        # getting response and save as response object
+        response = send_post_request(URL, APIKEY, SECRETKEY, 'stage', PHONENO, 'WAYSMS', message )
+        # print response if you want
+        print (response.text)
+    except Exception as e:
+        print(e)
```
Comment (jarifibrahim) [inline send_sms.py]: it would be nice to print a better message along with the exception.
How about 
```suggestion
        print("Failed to send SMS: %s" % e)
```

**Review by jarifibrahim — CHANGES_REQUESTED** [review]

File: main.py [hunk 9]
```diff
@@ -72,10 +71,11 @@ def main():
     files = get_file_list()
     new_files = get_files(files, DL_FILE_NAME)
     if new_files:
-        print 'Preparing email to send new files'
+        print ('Preparing email to send new files and Preparing sms to notify')
         prepare_email(new_files)
+        prepare_sms(new_files)
```
Comment (jarifibrahim) [inline main.py]: Can you make this optional, please?
Add a flag to config `NOTIFY_VIA_SMS` with `false` as default.
Then we can do 
```python
if NOTIFY_VIA_SMS:
        prepare_sms(new_files)
```

We should do this for the emails also https://github.com/jarifibrahim/resultBot/issues/6

File: config.py [hunk 10]
```diff
@@ -21,6 +21,12 @@
 PASSWORD = ''  # App password
 EMAIL_SMTP = 'smtp.gmail.com:587'
 
+# sms authentication credentials
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
+APIKEY = '' # way2sms apiKey
```
Comment (jarifibrahim) [inline config.py]: PEP8 naming conventions please :)
```suggestion
WAY2SMS_API_KEY = '' # way2sms apiKey
```

File: config.py [hunk 11]
```diff
@@ -21,6 +21,12 @@
...
+SECRETKEY = '' # way2sms secretKey
```
Comment (jarifibrahim) [inline config.py]:
```suggestion
WAY2SMS_SECRET_KEY = '' # way2sms secretKey
```

File: config.py [hunk 12]
```diff
@@ -21,6 +21,12 @@
 PASSWORD = ''  # App password
 EMAIL_SMTP = 'smtp.gmail.com:587'
 
+# sms authentication credentials
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
```
Comment (jarifibrahim) [inline config.py]:
```suggestion
WAY2SMS_URL = 'http://www.way2sms.com/api/v1/sendCampaign'
```

File: config.py [hunk 13]
```diff
@@ -21,6 +21,12 @@
...
+APIKEY = '' # way2sms apiKey
+SECRETKEY = '' # way2sms secretKey
+PHONENO = # receiver's valid phone number(indian) without country code
```
Comment (jarifibrahim) [inline config.py]:
```suggestion
PHONE_NO = # receiver's valid phone number(indian) without country code
```

**Review by jarifibrahim — APPROVED** [review]
Almost ready to merge :tada:

**Comment by jarifibrahim** [comment]
@nishikeshKardak Just to confirm, have you tried running the PR with way2sms configurations? Did you receive the SMS?

File: send_sms.py (hunk 7 above)
Comment (nishikeshKardak) [reply send_sms.py]: Hey @jarifibrahim if the SMS failed, then it would have stopped the program or not?

Comment (nishikeshKardak) [reply send_sms.py]: At this moment, I can't understand what to do if SMS failed. Can you help me?
If SMS failed then I would need to have exit from program or  have to continue?

Comment (jarifibrahim) [reply send_sms.py]: No. We should not kill the program if the sms failed. Let's keep the programming running even if the sms failed to send.

Comment (nishikeshKardak) [reply send_sms.py]: Hey @jarifibrahim I did following changes. Can you review this?

```
def send_sms(message):
    try:
        # get response
        response = send_post_request(WAY2SMS_URL, SMS_API_KEY, SMS_SECRET_KEY, 'stage', PHONE_NO, 'WAYSMS', message )
        # print response if you want
        res =  response.text
        if 'success' in res:
            print('\033[92m' + 'Successfully sent the sms' + '\033[0m')
    except requests.exceptions.HTTPError as eh:
        print('\033[91m' + 'An HTTP error occurred. %s' %eh)
        print('\033[0m') 
    except requests.exceptions.ConnectionError as ec:
        print('\033[91m' + 'A Connection error occurred %s' %ec)
        print('\033[0m')
    except requests.exceptions.Timeout as et:
        print('\033[91m' + 'It is ConnectTimeout or ReadTimeout %s' %et)
        print('\033[0m')
    except Exception as e:
        print('\033[91m' + "Something went wrong. Failed to send SMS: %s" % e)
        print('\033[0m')
    finally :
         pass
```

File: send_sms.py (hunk 6 above)
Comment (nishikeshKardak) [reply send_sms.py]: Sure we can remove this.

File: send_sms.py (hunk 7 above)
Comment (jarifibrahim) [reply send_sms.py]: @nishikeshKardak I think we don't need to handle all the exceptions. Look at this example
```python
>>> def try_and_print(req, params):
...  try:
...   requests.post(req, params)
...  except Exception as e:
...   print("Failed to send SMS: %s" % e)
>>> try_and_print("foo", "bar")
Failed to send SMS: Invalid URL 'foo': No schema supplied. Perhaps you meant http://foo?
>>> try_and_print("http:://google.com", "bar")
Failed to send SMS: No connection adapters were found for 'http:://google.com/'
>>> try_and_print("http://google.com", "bar")
>>> try_and_print("http://localhost", "bar")
Failed to send SMS: HTTPConnectionPool(host='localhost', port=80): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f89424932d0>: Failed to establish a new connection: [Errno 111] Connection refused',))
```

So a simple
```python
except Exception as e:
    print("Failed to send SMS: %s " % e)
```
should be enough

**Comment by nishikeshKardak** [comment]
> @nishikeshKardak Just to confirm, have you tried running the PR with way2sms configurations? Did you receive the SMS?

Yes,  I received sms with way2sms configuration. That's why I sent the PL.

**Comment by nishikeshKardak** [comment]
> @nishikeshKardak Just to confirm, have you tried running the PR with way2sms configurations? Did you receive the SMS?

Yes, I received sms with way2sms configuration.

**Comment by nishikeshKardak** [comment]
Hey @jarifibrahim  I did some changes. Can you review this??

File: send_sms.py (hunk 4 above)
Comment (jarifibrahim) [reply send_sms.py]: @nishikeshKardak clean up this one

**Comment by nishikeshKardak** [comment]
Hey @jarifibrahim Thanks for merging PL.

File: send_sms.py (hunk 4 above)
Comment (nishikeshKardak) [reply send_sms.py]: Referencing this #9
